src/System.Net.Security/tests/FunctionalTests/NegotiateStreamStreamToStreamTest.cs

```diff
@@ -0,0 +1,252 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Linq;
+using System.Security.Principal;
+using System.Text;
+using System.Threading.Tasks;
+
+using Xunit;
+
+namespace System.Net.Security.Tests
+{
+    public class NegotiateStreamStreamToStreamTest
+    {
+        private readonly byte[] _sampleMsg = Encoding.UTF8.GetBytes("Sample Test Message");
+
+        [Fact]
+        public void NegotiateStream_StreamToStream_Authentication_Success()
+        {
+            MockNetwork network = new MockNetwork();
+
+            using (var clientStream = new FakeNetworkStream(false, network))
+            using (var serverStream = new FakeNetworkStream(true, network))
+            using (var client = new NegotiateStream(clientStream))
+            using (var server = new NegotiateStream(serverStream))
+            {
+                Assert.False(client.IsAuthenticated);
+                Assert.False(server.IsAuthenticated);
+
+                Task[] auth = new Task[2];
+                auth[0] = client.AuthenticateAsClientAsync();
+                auth[1] = server.AuthenticateAsServerAsync();
+
+                bool finished = Task.WaitAll(auth, TestConfiguration.PassingTestTimeoutMilliseconds);
+                Assert.True(finished, "Handshake completed in the allotted time");
+
+                // Expected Client property values:
+                Assert.True(client.IsAuthenticated);
+                Assert.Equal(TokenImpersonationLevel.Identification, client.ImpersonationLevel);
+                Assert.Equal(true, client.IsEncrypted);
+                Assert.Equal(false, client.IsMutuallyAuthenticated);
+                Assert.Equal(false, client.IsServer);
+                Assert.Equal(true, client.IsSigned);
+                Assert.Equal(false, client.LeaveInnerStreamOpen);
+
+                IIdentity serverIdentity = client.RemoteIdentity;
+                Assert.Equal("NTLM", serverIdentity.AuthenticationType);
+                Assert.Equal(false, serverIdentity.IsAuthenticated);
+                Assert.Equal("", serverIdentity.Name);
+
+                // Expected Server property values:
+                Assert.True(server.IsAuthenticated);
+                Assert.Equal(TokenImpersonationLevel.Identification, server.ImpersonationLevel);
+                Assert.Equal(true, server.IsEncrypted);
+                Assert.Equal(false, server.IsMutuallyAuthenticated);
+                Assert.Equal(true, server.IsServer);
+                Assert.Equal(true, server.IsSigned);
+                Assert.Equal(false, server.LeaveInnerStreamOpen);
+
+                IIdentity clientIdentity = server.RemoteIdentity;
+                Assert.Equal("NTLM", clientIdentity.AuthenticationType);
+
+                Assert.Equal(true, clientIdentity.IsAuthenticated);
+                Assert.Equal(WindowsIdentity.GetCurrent().Name, clientIdentity.Name);
```

Comment:
Good point: I'll create a separate class to perform the clientIdentity name check. When the Unix part is ready we can use the same technique as with production code an include one or the other implementation of this class.